### PR TITLE
Moved most of cli.js into a new runner.js

### LINKED ISSUE #283: Split the CLI into a new class
As per https://github.com/lesshint/lesshint/issues/281, it would be useful to be able to call the CLI's logic via `require`/`import` rather than some form of subprocess.

Required for #278.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #283.

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
@shellscape is this what you meant by moving the essential bits into the main code?
@jwilsson is this design something you're ok with? *(you mentioned being up for discussing it)*

File: lib/runner.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+Runner.prototype.run = function () {
+    var lesshint = new Lesshint();
+    var exitDefer = Vow.defer();
+    var exitPromise = exitDefer.promise();
+    var promises = [];
+    var config;
+
+    if (!this.settings.args.length) {
+        console.error('No files to lint were passed. See lesshint -h');
```
Comment (jwilsson) [inline lib/runner.js]: Wouldn't it be better to reject the promise with a error message and a error code and keep all the `console.log`s in `cli.js`. So this file can be used from another Node program without any logging output?

File: lib/runner.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        if (this.settings.exclude) {
+            config.excludedFiles.push(this.settings.exclude);
+        }
+
+        if (this.settings.linters) {
+            config.linters.push.apply(config.linters, this.settings.linters);
+        }
+    } catch (e) {
+        console.error("Something's wrong with the config file. Error: " + e.message);
```
Comment (jwilsson) [inline lib/runner.js]: Same reasoning as above.

File: lib/runner.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        var warnings;
+
+        results = [].concat.apply([], results);
+
+        reporter = lesshint.getReporter(this.settings.reporter);
+
+        if (reporter) {
+            reporter.report(results);
+        } else {
+            console.error('Could not find reporter "%s".', this.settings.reporter);
```
Comment (jwilsson) [inline lib/runner.js]: Same reasoning as above.

File: lib/runner.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            }, 0);
+
+            if (warnings > maxWarnings) {
+                exitDefer.reject(EXIT_CODES.WARNING);
+            } else {
+                exitDefer.resolve(EXIT_CODES.OK);
+            }
+        }
+    }, this).fail(function (error) {
+        console.error('An unknown error occurred when checking "%s", please file an issue with this info:', error.lesshintFile);
```
Comment (jwilsson) [inline lib/runner.js]: Same reasoning as above.

**Comment by jwilsson** [comment]
I have some thoughts about logging from inside the runner but other than that I'm good.

File: lib/runner.js (hunk 1 above)
Comment (JoshuaKGoldberg) [inline lib/runner.js]: Agreed! Will change.

**Comment by JoshuaKGoldberg** [comment]
The `'should exit without errors when passed a built-in reporter name'` spec test suddenly failing confused me for a while... but it looks like it's not actually an accurate test anymore?

> ```javascript
> result = cli({
>    args: [path.join(__dirname, '../data/files/ok.less')],
>    reporter: 'stylish'
> });
>
> return result.then(function (status) {
>     expect(status).to.equal(0);
> });
> ```

`'stylish'` isn't included with `lesshint`. The test asserts that the returned status is 0, but doesn't care that `console.error` was called during `lesshint` with `Could not find reporter "stylish".`. 

> ```javascript
> reporter = lesshint.getReporter(program.reporter);
>
> if (reporter) {
>     reporter.report(results);
> } else {
>     console.error('Could not find reporter "%s".', program.reporter);
> }
> ```

When `runner.js` started resolving and returning, the behavior changed to now correctly give an error code.

**Comment by JoshuaKGoldberg** [comment]
Another change proposal for the CLI tests: `result.then(...)` instead of `result.fail(...)`.

**Comment by jwilsson** [comment]
> Question from this: are you ok with no longer giving a 0 error status on an invalid linter name?
Yes, that seems like the correct behaviour.

> Another change proposal for the CLI tests: `result.then(...)` instead of `result.fail(...)`. 
Wouldn't it be better to separate the `then()` and `catch()` in `cli.js` instead? Cause' I'm guessing that's why, right? But still use `always()` for the exit code for example.

**Comment by JoshuaKGoldberg** [comment]
Ok. Bad news, good news.

Bad news: it's been a month since you last commented. Sorry about that. PR updated & merge conflicts resolved.

Good news: I reverted the functional changes to the CLI such that it again resolves or rejects as appropriate. I *also* became very irritated with how the CLI failing to resolve or reject appropriately would cause errors with any subsequent test, so I made the runner and its constituent reporters take in a "`logger`" object (à la dependency injection). The gist of it is:
* `bin/lesshint` doesn't pass `exit` to `cli`; instead, it retrieves the result and calls `exit` with it.
* `lib/cli.js` takes in a `logger` instead of directly using `console` and passes that to its `Runner`.
* `lib/runner.js::Runner` takes in a `logger` and passes it to its reporters.
* reporters are now given a `logger` object after `errors` that they can (should) use to log.

Using a dependency injected `logger` gives the benefit that consumers of the reporters (such as our tests or external consumers of `lesshint`) don't need to globally mock or spy or stub the global console.

One downside is that existing custom linters won't be aware of the change, but since it's backwards compatible (doesn't change behavior for existing consumers) I don't think that should be a problem.

Thoughts?

**Comment by shellscape** [comment]
> don't need to globally mock or spy or stub the global console.

In what situation would they ever have to mock/spy/stub `console.log`?

File: lib/lesshint.js [hunk 5]
```diff
@@ -133,7 +133,7 @@ Lesshint.prototype.getReporter = function (reporter) {
     var reporterPath;
 
     // Nothing defined, just fall back to our simple default
-    if (!reporter) {
+    if (!reporter || reporter === 'default') {
```
Comment (shellscape) [inline lib/lesshint.js]: cherry pick this as a separate bug fix

**Comment by JoshuaKGoldberg** [comment]
> In what situation would they ever have to mock/spy/stub console.log?

Generally any utility that would want to call the runner as a part of some other library. I'm thinking of [autolesshint](https://github.com/automutate/autolesshint) specifically.

```javascript
var cli = require('lesshint/lib/cli');

module.exports = function (logger) {
    return cli(
        {
            args: ['index.less'],
            config: '.lesshintrc'
        },
        logger);
};
```

> This is starting to look like a classic case of trying to do too much in a single PR.

Fair. The vision has expanded in scope from "split CLI into CLI and Runner" into "split CLI into CLI and *a bug-free when in standalone* Runner". Filing issues now...

**Comment by jwilsson** [comment]
It looks like this PR has started to go off-track. Perhaps it's better to file a new one once we ironed out #281 and #300? Just my two-cents.

**Comment by JoshuaKGoldberg** [comment]
:+1: will keep code around for easy viewing

**Pull request closed without merging**